Cabal/Distribution/Simple/Configure.hs

```diff
@@ -1036,6 +1007,62 @@ configureCoverage verbosity cfg comp = do
            ++ "program coverage. Program coverage has been disabled.")
         return id
 
+-- | Select and apply profiling settings for the build based on the
+-- 'ConfigFlags' and 'Compiler'.
+configureProfiling :: Verbosity -> ConfigFlags -> Compiler
+                   -> IO (LocalBuildInfo -> LocalBuildInfo)
+configureProfiling verbosity cfg comp = do
+  -- The --profiling flag sets the default for both libs and exes,
+  -- but can be overidden by --library-profiling, or the old deprecated
+  -- --executable-profiling flag.
+  --
+  -- The --profiling-detail and --library-profiling-detail flags behave
+  -- similarly
+  let tryExeProfiling = fromFlagOrDefault False
+                        (mappend (configProf cfg) (configProfExe cfg))
+      tryLibProfiling = fromFlagOrDefault tryExeProfiling
+                        (mappend (configProf cfg) (configProfLib cfg))
+
+      tryExeProfileLevel = fromFlagOrDefault ProfDetailDefault
+                           (configProfDetail cfg)
+      tryLibProfileLevel = fromFlagOrDefault ProfDetailDefault
+                           (mappend
+                            (configProfDetail cfg)
+                            (configProfLibDetail cfg))
+
+      checkProfileLevel (ProfDetailOther other) = do
+        warn verbosity
+          ("Unknown profiling detail level '" ++ other
+           ++ "', using default.\nThe profiling detail levels are: "
+           ++ intercalate ", "
+           [ name | (name, _, _) <- knownProfDetailLevels ])
+        return ProfDetailDefault
+      checkProfileLevel other = return other
+
+  (profEnabled, applyProfiling) <-
+    if profilingSupported comp
+    then do
+      exeLevel <- checkProfileLevel tryExeProfileLevel
+      libLevel <- checkProfileLevel tryLibProfileLevel
+      let apply lbi = lbi { withProfLib       = tryLibProfiling
+                          , withProfLibDetail = libLevel
+                          , withProfExe       = tryExeProfiling
+                          , withProfExeDetail = exeLevel
+                          }
+      return (tryExeProfiling || tryLibProfiling, apply)
+    else do
+      when (tryExeProfiling || tryLibProfiling) $ warn verbosity
+        ("The compiler " ++ showCompilerId comp ++ " does not support "
+         ++ "profiling. Profiling has been disabled.")
+      return (False, id)
+
+  when profEnabled $ warn verbosity
```

Comment:
Hmm, is this correct? Was: `when (profEnabledExe && not profEnabledLib)`. Now: `when (tryExeProfiling || tryLibProfiling)`.